### ISSUE TITLE
Model improves worse with GradientTape than with fit()

### ISSUE THREAD

**Issue by matteopilz** [tensorflow/tensorflow#57392]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

tf 2.9.1

### Custom Code

Yes

### OS Platform and Distribution

Linux Ubuntu 22.04.1

### Python version

3.8 & 3.10

### Current Behaviour?

```shell
I already opened a question on stackoverflow for this, but I believe it could be some issue with tensorflow.

https://stackoverflow.com/questions/73440964/model-not-improving-with-gradienttape-but-with-model-fit

The same model using either custom training with GradientTape or keras' model.fit() will perform differently. The custom run, will stop improving earlier and will take longer to improve at all.

I've also included a Google Colab notebook in order to reproduce the issue:
https://colab.research.google.com/drive/1pk66rbiux5vHZcav9VNSBhdWWIhQM-nF?usp=sharing
```

### Standalone code to reproduce the issue

```shell
def build_model(kernel_regularizer=l2(0.0001), dropout=0.001, recurrent_dropout=0.):
    x1 = Input(62)
    x2 = Input((62, 3))

    x = Embedding(30, 100, mask_zero=True)(x1)
    x = Concatenate()([x, x2])

    x = Bidirectional(LSTM(500,
                           return_sequences=True,
                           kernel_regularizer=kernel_regularizer,
                           dropout=dropout,
                           recurrent_dropout=recurrent_dropout))(x)

    x = Bidirectional(LSTM(500,
                           return_sequences=False,
                           kernel_regularizer=kernel_regularizer,
                           dropout=dropout,
                           recurrent_dropout=recurrent_dropout))(x)

    x = Activation('softmax')(x)

    x = Dense(1000)(x)
    x = Dense(500)(x)
    x = Dense(250)(x)
    x = Dense(1, bias_initializer='ones')(x)

    x = tf.math.abs(x)
    return Model(inputs=[x1, x2], outputs=x)


optimizer = Adam(learning_rate=0.0001)

model = build_model()
model.compile(optimizer=optimizer, loss='mse', metrics='mse')

options = tf.data.Options()
options.experimental_distribute.auto_shard_policy = AutoShardPolicy.DATA 
dat_train = tf.data.Dataset.from_generator(
    generator= lambda: <load_function()> 
    output_types=((tf.int32, tf.float32), tf.float32)
) 
dat_train = dat_train.with_options(options) 

# keras training
model.fit(dat_train, epochs=50)


# custom training
for epoch in range(50):
    for (x1, x2), y in dat_train:
        with tf.GradientTape() as tape:
            y_pred = model((x1, x2), training=True)
            loss = model.loss(y, y_pred)
        grads = tape.gradient(loss, model.trainable_variables)
        model.optimizer.apply_gradients(zip(grads, model.trainable_variables))
```

### Relevant log output

_No response_</details>

**Comment by matteopilz** [comment]
I have found the solution, it was a simple shape mismatch, which was somehow not picked up by any error check and worked both with my custom loss function and MSE. Using `x = Reshape(())(x)` as final layer did the trick.
